### ISSUE TITLE
AWS - unable to deploy OCP infra on European Zone or on default us-east-1 zone.

### ISSUE THREAD

**Issue by dwojciec** [openshift/openshift-ansible-contrib#142]
I tried to deploy OCP on AWS using the python script :
./ose-on-aws.py --keypair=OSE-key --region eu-central-1 --ami ami-7def1712 --create-key=yes --key-path=/root/.ssh/id_rsa.pub --rhsm-user=dwojciec --rhsm-password=password \
> --public-hosted-zone=mydomain.com --rhsm-pool="Red Hat OpenShift Container Platform, Standard, 2-Core" \
> --github-client-secret=6746b9659154d680a06ee9ba07b1d379582ab40f --github-organization=myorg-openshift \
>  --github-client-id=f64169d9eaa3efccf87e -vvv

The error received is : 
TASK [cloudformation-infra : Create Greenfield Infrastructure] *****************
task path: /root/AWS/openshift-ansible-contrib/reference-architecture/aws-ansible/playbooks/roles/cloudformation-infra/tasks/main.yaml:2
Using module file /usr/lib/python2.7/site-packages/ansible/modules/core/cloud/amazon/cloudformation.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145 `" && echo ansible-tmp-1483376807.65-246631564182145="` echo $HOME/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmph8KP_7 TO /root/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145/cloudformation.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145/ /root/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145/cloudformation.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145/cloudformation.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1483376807.65-246631564182145/" > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "disable_rollback": false, 
            "ec2_url": null, 
            "notification_arns": null, 
            "profile": null, 
            "region": "eu-central-1", 
            "security_token": null, 
            "stack_name": "openshift-infra", 
            "stack_policy": null, 
            "state": "present", 
            "tags": null, 
            "template": "roles/cloudformation-infra/files/greenfield.json", 
            "template_format": null, 
            "template_parameters": {
                "AmiId": "ami-7def1712", 
                "AppNodeInstanceType": "t2.medium", 
                "AppWildcardDomain": "*.apps.dwojciec.com", 
                "BastionInstanceType": "t2.micro", 
                "BastionRootVolType": "gp2", 
                "BastionUserData": "I2Nsb3VkLWNvbmZpZwp1c2VyczoKLSBkZWZhdWx0CgpzeXN0ZW1faW5mbzoKICBkZWZhdWx0X3VzZXI6CiAgICBuYW1lOiBlYzItdXNlcg==", 
                "InfraDockerVolSize": 25, 
                "InfraDockerVolType": "gp2", 
                "InfraInstanceType": "t2.medium", 
                "InfraRootVolSize": 15, 
                "InfraRootVolType": "gp2", 
                "KeyName": "OSE-key", 
                "MasterApiPort": "443", 
                "MasterClusterHostname": "internal-openshift-master.dwojciec.com", 
                "MasterClusterPublicHostname": "openshift-master.dwojciec.com", 
                "MasterDockerVolSize": 25, 
                "MasterDockerVolType": "gp2", 
                "MasterEmptyVolSize": 5, 
                "MasterEmptyVolType": "gp2", 
                "MasterEtcdVolSize": 25, 
                "MasterEtcdVolType": "gp2", 
                "MasterHealthTarget": "TCP:443", 
                "MasterInstanceType": "m4.large", 
                "MasterRootVolSize": 10, 
                "MasterRootVolType": "gp2", 
                "MasterUserData": "I2Nsb3VkLWNvbmZpZwpjbG91ZF9jb25maWdfbW9kdWxlczoKLSBkaXNrX3NldHVwCi0gbW91bnRzCgpmc19zZXR1cDoKLSBsYWJlbDogZXRjZF9zdG9yYWdlCiAgZmlsZXN5c3RlbTogeGZzCiAgZGV2aWNlOiAvZGV2L3h2ZGMKICBwYXJ0aXRpb246IGF1dG8KLSBsYWJlbDogZW1wdHlkaXIKICBmaWxlc3lzdGVtOiB4ZnMKICBkZXZpY2U6IC9kZXYveHZkZAogIHBhcnRpdGlvbjogYXV0bwoKcnVuY21kOgotIG1rZGlyIC1wIC92YXIvbGliL2V0Y2QKLSBta2RpciAtcCAvdmFyL2xpYi9vcmlnaW4vb3BlbnNoaWZ0LmxvY2FsLnZvbHVtZXMKCm1vdW50czoKLSBbIC9kZXYveHZkYywgL3Zhci9saWIvZXRjZCwgeGZzLCAiZGVmYXVsdHMiIF0KLSBbIC9kZXYveHZkZCwgL3Zhci9saWIvb3JpZ2luL29wZW5zaGlmdC5sb2NhbC52b2x1bWVzLCB4ZnMsICJkZWZhdWx0cyxncXVvdGEiIF0KCgp3cml0ZV9maWxlczoKLSBjb250ZW50OiB8CiAgICBERVZTPScvZGV2L3h2ZGInCiAgICBWRz1kb2NrZXJfdm9sCiAgICBEQVRBX1NJWkU9OTUlVkcKICAgIEVYVFJBX0RPQ0tFUl9TVE9SQUdFX09QVElPTlM9Ii0tc3RvcmFnZS1vcHQgZG0uYmFzZXNpemU9M0ciCiAgcGF0aDogL2V0Yy9zeXNjb25maWcvZG9ja2VyLXN0b3JhZ2Utc2V0dXAKICBvd25lcjogcm9vdDpyb290Cgp1c2VyczoKLSBkZWZhdWx0CgpzeXN0ZW1faW5mbzoKICBkZWZhdWx0X3VzZXI6CiAgICBuYW1lOiBlYzItdXNlcg==", 
                "NodeDockerVolSize": 25, 
                "NodeDockerVolType": "gp2", 
                "NodeEmptyVolSize": 50, 
                "NodeEmptyVolType": "gp2", 
                "NodeRootVolSize": 15, 
                "NodeRootVolType": "gp2", 
                "NodeUserData": "I2Nsb3VkLWNvbmZpZwpjbG91ZF9jb25maWdfbW9kdWxlczoKLSBkaXNrX3NldHVwCi0gbW91bnRzCgpmc19zZXR1cDoKLSBsYWJlbDogZW1wdHlkaXIKICBmaWxlc3lzdGVtOiB4ZnMKICBkZXZpY2U6IC9kZXYveHZkYwogIHBhcnRpdGlvbjogYXV0bwoKcnVuY21kOgotIG1rZGlyIC1wIC92YXIvbGliL29yaWdpbi9vcGVuc2hpZnQubG9jYWwudm9sdW1lcwoKbW91bnRzOgotIFsgL2Rldi94dmRjLCAvdmFyL2xpYi9vcmlnaW4vb3BlbnNoaWZ0LmxvY2FsLnZvbHVtZXMsIHhmcywgImRlZmF1bHRzLGdxdW90YSIgXQoKd3JpdGVfZmlsZXM6Ci0gY29udGVudDogfAogICAgREVWUz0nL2Rldi94dmRiJwogICAgVkc9ZG9ja2VyX3ZvbAogICAgREFUQV9TSVpFPTk1JVZHCiAgICBFWFRSQV9ET0NLRVJfU1RPUkFHRV9PUFRJT05TPSItLXN0b3JhZ2Utb3B0IGRtLmJhc2VzaXplPTNHIgogIHBhdGg6IC9ldGMvc3lzY29uZmlnL2RvY2tlci1zdG9yYWdlLXNldHVwCiAgb3duZXI6IHJvb3Q6cm9vdAoKdXNlcnM6Ci0gZGVmYXVsdAoKc3lzdGVtX2luZm86CiAgZGVmYXVsdF91c2VyOgogICAgbmFtZTogZWMyLXVzZXI=", 
                "PublicHostedZone": "dwojciec.com", 
                "Region": "eu-central-1", 
                "Route53HostedZone": "dwojciec.com.", 
                "S3BucketName": "openshift-infra-ocp-registry-dwojciec", 
                "S3User": "openshift-infra-s3-openshift-user", 
                "SubnetAvailabilityZones": "eu-central-1a,eu-central-1b", 
                "SubnetCidrBlocks": "10.20.1.0/24,10.20.2.0/24,10.20.3.0/24,10.20.4.0/24,10.20.5.0/24,10.20.6.0/24", 
                "VpcCidrBlock": "10.20.0.0/16", 
                "VpcName": "ose-multi-az-vpc-openshift-infra"
            }, 
            "template_url": null, 
            "validate_certs": true
        }, 
        "module_name": "cloudformation"
    }
}

MSG:

Template error: Fn::Select  cannot select nonexistent value at index 2

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=1   

I tried using the default availability zone US-EAST-1  instead of European zone and I received this issue inside cloudformation  console:
CREATE_FAILED  AWS::EC2::Subnet  PublicSubnet3  Value (us-east-1c) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: us-east-1e, us-east-1a, us-east-1b, us-east-1d.

I only succeeded by using : --region us-east-2 --ami ami-0a33696f

**Comment by detiber** [comment]
@cooktheryan are the zones for us-east-1 still hardcoded? That may be the issue for us-east-1. For European zones, I know some have only 2 AZs.

**Comment by cooktheryan** [comment]
Yes, we still hard code the expectation that there are at least 3 AZs. That definitely explains the issue with the European zones.

**Comment by dwojciec** [comment]
After cleaning my eu-west-1 zone. I deleted my VM I used to provision the infrastructure and I decided to use a Vbox image coming from a VagrantFile. Once VPC, all VM cleaned on eu-west-1 I was able to provision an OCP env on eu-west-1 using this script :
./ose-on-aws.py  -v \
	--public-hosted-zone=dwojciec.com \
	--keypair=OSE-key \
	--rhsm-user=USER --rhsm-password=PASSWORD \
	--rhsm-pool="8a85f9815695736e015697c1490a0ff" \
	--master-instance-type=m4.large \
	--node-instance-type=m4.xlarge \
	--app-instance-type=m4.xlarge \
	--region=eu-west-1 \
	--ami=ami-8b8c57f8 \
	--create-key=yes \
	--key-path=/root/.ssh/id_rsa.pub \
	--github-client-secret=6746b9659154d680a06ee9ba07b1d379582ab50f\
	--github-organization=dwojciec-openshift \
   --github-client-id=f64169d9eaa3efcc67e

For the US zone like us-east-1 I received the same error as yesterday  during the cloudformation creation step :
Value (us-east-1c) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: us-east-1e, us-east-1a, us-east-1b, us-east-1d. 
Seems due to my AWS account. It's too old. I tried to create a VPC using AWS UI and I can't use us-east-1c AZ. The good news is : the problem is not the python script ;)
